File: src/main/java/club/xiaojiawei/bean/area/PlayArea.java
package club.xiaojiawei.bean.area;

import club.xiaojiawei.bean.entity.Card;
import club.xiaojiawei.enums.HeroTypeEnum;
import club.xiaojiawei.enums.PowerTypeEnum;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import lombok.extern.slf4j.Slf4j;

import java.util.Objects;

import static club.xiaojiawei.enums.CardTypeEnum.*;

/**
 * 战场
 * @author 肖嘉威
 * @date 2022/11/27 15:02
 */
@Getter
@Setter
@ToString
@Slf4j
public class PlayArea extends Area {

    private HeroTypeEnum heroType;
    private PowerTypeEnum powerType;
    private Card hero;
    private Card heroHide;
    private Card power;
    private Card powerHide;
    private Card weapon;
    private Card weaponHide;

    @Override
    public boolean add(Card card, int pos){
        boolean result = true;
        if (card == null){
            result = false;
        }else if (card.getCardType() == HERO_POWER){
            powerHide = power;
            power = card;
//            TODO 设置powerType
            addZoneAndLog("技能", card);
        }else if (card.getCardType() == HERO){
//            TODO 设置heroType
            heroHide = hero;
            hero = card;
            addZoneAndLog("英雄", card);
        }else if (card.getCardType() == WEAPON){
            weaponHide = weapon;
            weapon = card;
            addZoneAndLog("武器", card);
        }else {
            result =  super.add(card, pos);
        }
        return result;
    }

    private void addZoneAndLog(String name, Card card){
        addZone(card);
        logInfo(card, name);
    }

    @Override
    public Card findByEntityId(String entityId) {
        Card card = super.findByEntityId(entityId);
        if (card == null){
            if (hero != null && Objects.equals(hero.getEntityId(), entityId)){
                card = hero;
            }else if (power != null && Objects.equals(power.getEntityId(), entityId)){
                card = power;
            }else if (weapon != null && Objects.equals(weapon.getEntityId(), entityId)){
                card = weapon;
            }else if (heroHide != null && Objects.equals(heroHide.getEntityId(), entityId)){
                card = heroHide;
            }else if (powerHide != null && Objects.equals(powerHide.getEntityId(), entityId)){
                card = powerHide;
            }else if (weaponHide != null && Objects.equals(weaponHide.getEntityId(), entityId)){
                card = weaponHide;
            }
        }
        return card;
    }


    @Override
    public Card removeByEntityId(String entityId) {
        Card card = super.removeByEntityId(entityId);
        if (card == null){
            if (hero != null && Objects.equals(hero.getEntityId(), entityId)){
                card = hero;
                hero = null;
            }else if (power != null && Objects.equals(power.getEntityId(), entityId)){
                card = power;
                power = null;
            }else if (weapon != null && Objects.equals(weapon.getEntityId(), entityId)){
                card = weapon;
                weapon = null;
            }else if (heroHide != null && Objects.equals(heroHide.getEntityId(), entityId)){
                card = heroHide;
                heroHide = null;
            }else if (powerHide != null && Objects.equals(powerHide.getEntityId(), entityId)){
                card = powerHide;
                powerHide = null;
            }else if (weaponHide != null && Objects.equals(weaponHide.getEntityId(), entityId)){
                card = weaponHide;
                weaponHide = null;
            }
        }
        return card;
    }

    public PlayArea() {
        super(7);
    }

}

File: Hearthstone-Script-Card-SDK/src/main/kotlin/club/xiaojiawei/bean/area/PlayArea.kt
package club.xiaojiawei.bean.area

import club.xiaojiawei.bean.Card
import club.xiaojiawei.bean.Player
import club.xiaojiawei.config.ENABLE_AREA_LOG
import club.xiaojiawei.config.log
import club.xiaojiawei.enums.CardTypeEnum
import club.xiaojiawei.util.isTrue

/**
 * 战场
 * @author 肖嘉威
 * @date 2022/11/27 15:02
 */
class PlayArea(
    allowLog: Boolean = false,
    player: Player,
) : Area(allowLog = allowLog, maxSize = 7, player = player) {
    @Volatile
    var hero: Card? = null

    @Volatile
    var heroHide: Card? = null

    @Volatile
    var power: Card? = null

    @Volatile
    var powerHide: Card? = null

    @Volatile
    var weapon: Card? = null

    @Volatile
    var weaponHide: Card? = null

    private fun addZoneAndLog(
        name: String,
        card: Card,
    ) {
        addZone(card)

        (allowLog && ENABLE_AREA_LOG).isTrue {
            log.info { getLogText(card, name) }
        }
    }

    /**
     * 获取当前法强
     */
    fun getSpellPower(): Int =
        cards.sumOf { card -> card.spellPower } +
                (hero?.spellPower ?: 0) + (power?.spellPower ?: 0) + (weapon?.spellPower ?: 0)

    override fun removeCard(index: Int): Card? {
        val removedCard = super.removeCard(index)
        removedCard?.action?.triggerRemovedToPlayArea(player.war)
        return removedCard
    }

    override fun add(
        card: Card?,
        pos: Int,
    ): Boolean {
        card?.let {
            if (pos > 0) {
                val copyPlayCards = player.playArea.cards.toList()
                for (card in copyPlayCards) {
                    card.action.triggerAddCardToMyPlayArea(player.war, it)
                }
                val copyHandCards = player.handArea.cards.toList()
                for (card in copyHandCards) {
                    card.action.triggerAddCardToMyPlayArea(player.war, it)
                }
                card.action.triggerBattlecry(player.war)
            }
            card.action.triggerAddedToPlayArea(player.war)
        }
        var result = true
        if (card == null) {
            result = false
        } else if (card.cardType === CardTypeEnum.HERO_POWER) {
            powerHide = power
            power = card
            addZoneAndLog("技能", card)
        } else if (card.cardType === CardTypeEnum.HERO) {
            heroHide = hero
            hero = card
            addZoneAndLog("英雄", card)
        } else if (card.cardType === CardTypeEnum.WEAPON) {
            weaponHide = weapon
            weapon = card
            addZoneAndLog("武器", card)
        } else {
            result = super.add(card, pos)
        }
        return result
    }

    override fun findByEntityId(entityId: String): Card? {
        var card = super.findByEntityId(entityId)
        if (card == null) {
            if (hero?.entityId == entityId) {
                card = hero
            } else if (power?.entityId == entityId) {
                card = power
            } else if (weapon?.entityId == entityId) {
                card = weapon
            } else if (heroHide?.entityId == entityId) {
                card = heroHide
            } else if (powerHide?.entityId == entityId) {
                card = powerHide
            } else if (weaponHide?.entityId == entityId) {
                card = weaponHide
            }
        }
        return card
    }

    override fun removeByEntityId(entityId: String): Card? {
        var card = super.removeByEntityId(entityId)
        if (card == null) {
            if (hero?.entityId == entityId) {
                card = hero
                hero = null
            } else if (power?.entityId == entityId) {
                card = power
                power = null
            } else if (weapon?.entityId == entityId) {
                card = weapon
                weapon = null
            } else if (heroHide?.entityId == entityId) {
                card = heroHide
                heroHide = null
            } else if (powerHide?.entityId == entityId) {
                card = powerHide
                powerHide = null
            } else if (weaponHide?.entityId == entityId) {
                card = weaponHide
                weaponHide = null
            }
        }
        return card
    }
}